File: app/src/main/java/com/jrvermeer/psalter/UI/Adaptors/PsalterPagerAdapter.java
package com.jrvermeer.psalter.UI.Adaptors;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.v4.view.PagerAdapter;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.jrvermeer.psalter.Core.Contracts.IPagerCallbacks;
import com.jrvermeer.psalter.Core.Contracts.IPsalterRepository;
import com.jrvermeer.psalter.Core.Models.Psalter;
import com.jrvermeer.psalter.Core.Util;
import com.jrvermeer.psalter.Infrastructure.PsalterDb;
import com.jrvermeer.psalter.R;

import java.security.spec.InvalidParameterSpecException;

/**
 * Created by Jonathan on 3/27/2017.
 */

public class PsalterPagerAdapter extends PagerAdapter {
    private Context _context;
    private IPsalterRepository psalterRepository;
    private IPagerCallbacks callbacks;
    private boolean nightMode;
    private boolean showScore = false;

    public PsalterPagerAdapter(Context context, IPsalterRepository repo, IPagerCallbacks callbacks, boolean showScore, boolean nightMode){
        _context = context;
        psalterRepository = repo;
        this.showScore = showScore;
        this.nightMode = nightMode;
        this.callbacks = callbacks;
    }

    @Override
    public Object instantiateItem(ViewGroup collection, int position){
        try{
             Psalter psalter = psalterRepository.getIndex(position);

            LayoutInflater inflater = LayoutInflater.from(_context);
            ViewGroup layout = (ViewGroup)inflater.inflate(R.layout.psalter_layout, collection, false);

            ((TextView)layout.findViewById(R.id.tvPagerHeading)).setText(psalter.getHeading());
            String lyrics = psalter.getLyrics();
            TextView tvLyrics = ((TextView)layout.findViewById(R.id.tvPagerLyrics));
            if(showScore){
                Drawable score = psalterRepository.getScore(psalter);
                if(score != null){
                    if(nightMode) Util.invertColors(score);
                    ((ImageView)layout.findViewById(R.id.imgScore)).setImageDrawable(score);
                    int lyricStartIndex = lyrics.indexOf((psalter.getNumVersesInsideStaff() + 1) + ". ");
                    if(lyricStartIndex > 0) tvLyrics.setText(lyrics.substring(lyricStartIndex));
                }
                else tvLyrics.setText(lyrics);
            }
            else tvLyrics.setText(lyrics);

            TextView tvPagerPsalm = layout.findViewById(R.id.tvPagerPsalm);
            tvPagerPsalm.setText(Html.fromHtml(psalter.getSubtitleLink()));
            tvPagerPsalm.setMovementMethod(LinkMovementMethod.getInstance());
            collection.addView(layout);

            callbacks.pageCreated(layout, position);
            return layout;
        } catch (Exception ex){
            return null;
        }
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        if(object != null){
            container.removeView((View)object);
        }
    }

    @Override
    public int getCount() {
        return psalterRepository.getCount();
    }

    @Override
    public CharSequence getPageTitle(int position) {
        try{
            return psalterRepository.getIndex(position).getTitle();
        } catch (Exception ex){
            return "?";
        }
    }


    public void showScore(){
        if(!showScore){
            showScore = true;
            notifyDataSetChanged();
        }
    }
    public void hideScore(){
        if(showScore){
            showScore = false;
            notifyDataSetChanged();
        }
    }


    @Override
    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }
}

File: app/src/main/java/com/jrvermeer/psalter/ui/adaptors/PsalterPagerAdapter.kt
package com.jrvermeer.psalter.ui.adaptors

import android.content.Context
import androidx.core.text.HtmlCompat
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.jrvermeer.psalter.models.Psalter
import com.jrvermeer.psalter.infrastructure.Logger
import com.jrvermeer.psalter.infrastructure.PsalterDb
import com.jrvermeer.psalter.*
import com.jrvermeer.psalter.helpers.DownloadHelper
import kotlinx.android.synthetic.main.psalter_layout.view.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import java.util.concurrent.ConcurrentHashMap

/**
 * Created by Jonathan on 3/27/2017.
 */

class PsalterPagerAdapter(private val context: Context,
                          private val scope: CoroutineScope,
                          private val psalterDb: PsalterDb,
                          private val downloader: DownloadHelper,
                          private var showScore: Boolean,
                          private val nightMode: Boolean) : androidx.viewpager.widget.PagerAdapter() {

    private val views = ConcurrentHashMap<Int, View>()

    fun getView(i: Int): View? {
        return views[i]
    }

    override fun instantiateItem(collection: ViewGroup, position: Int): Any {
        try {
            val psalter = psalterDb.getIndex(position)!!
            Logger.d("Viewpager building page for psalter ${psalter.title}")
            val inflater = LayoutInflater.from(context)
            val layout = inflater.inflate(R.layout.psalter_layout, collection, false) as ViewGroup

            scope.launch { setScoreAndLyrics(psalter, layout) }

            layout.tvPagerHeading.text = psalter.heading
            layout.tvPagerPsalm.text = HtmlCompat.fromHtml(psalter.bibleLink, HtmlCompat.FROM_HTML_MODE_LEGACY)
            layout.tvPagerPsalm.movementMethod = LinkMovementMethod.getInstance()

            views[position] = layout
            collection.addView(layout)

            return layout
        }
        catch (ex: Exception) {
            Logger.e("Error in PsalterPagerAdapter.instantiateItem", ex)
            return collection
        }
    }

    override fun destroyItem(container: ViewGroup, position: Int, item: Any) {
        if (item is View) {
            container.removeView(item)
            views.remove(position)
        }
    }

    override fun getCount(): Int {
        return psalterDb.getCount()
    }

    override fun getPageTitle(position: Int): CharSequence? {
        return psalterDb.getIndex(position)?.title ?: "?"
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }

    private suspend fun setScoreAndLyrics(psalter: Psalter, layout: View) {
        // load audio in bg. viewpager never needs audio, but user could request it at tap of a button
        scope.launch { psalter.loadAudio(downloader) }
        var text = psalter.lyrics
        if (showScore) {
            layout.scoreProgress.show()
            val score = psalter.loadScore(downloader)
            layout.scoreProgress.hide()
            if (score != null) {
                if (nightMode) score.invertColors()
                layout.imgScore.setImageDrawable(score)

                val lyricStartIndex = psalter.lyrics.indexOf((psalter.numVersesInsideStaff + 1).toString() + ". ")
                text = if (lyricStartIndex < 0) "" else psalter.lyrics.substring(lyricStartIndex)
            }
        }
        else {
            layout.imgScore.setImageDrawable(null)
            // load score in bg. viewpager doesn't need it *now*, but could at the tap of a button
            scope.launch { psalter.loadScore(downloader) }
        }
        layout.tvPagerLyrics.text = text
    }

    fun toggleScore()  {
        showScore = !showScore
        for((i, view) in views) {
            scope.launch { setScoreAndLyrics(psalterDb.getIndex(i)!!, view) }
        }
    }


}
